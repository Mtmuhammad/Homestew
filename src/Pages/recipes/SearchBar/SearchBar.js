import { React, useState, useEffect } from "react";
import "./SearchBar.scss";
import client from "../../../http-common";
import SmallPost from "../../../Components/SmallPost/SmallPost";
import Spinner from "../../../Components/Spinner/Spinner";


const SearchBar = () => {
  const [query, setQuery] = useState("");
  const [value, setValue] = useState("");
  const [recipes, setRecipes] = useState([]);
  const [randomRecipes, setRandomRecipes] = useState([]);
  const [isLoading, setIsLoading] = useState(false);
  const [offset, setOffset] = useState(56);

  useEffect(() => {
    setIsLoading(true);
    // parse local storage for saved random recipes
    let result = JSON.parse(localStorage.getItem("random"));

    //if found, add to state; else make api call, add results to state and local storage
    if (result) {
      setRandomRecipes(result);
      setIsLoading(false);
    } else {
      const getRandomRecipes = () => {
        setIsLoading(true);
        client
          .get("random", {
            params: {
              apiKey: process.env.REACT_APP_API_KEY4,
              number: 56,
            },
          })
          .then(function (res) {
            setRandomRecipes(res.data.recipes);
            setIsLoading(false);
            localStorage.setItem("random", JSON.stringify(res.data.recipes));
          })
          .catch(function (error) {
            console.log(error);
          });
      };

      getRandomRecipes();
    }
  }, []);

  const getRecipes = () => {
    // removes any recipe list 
    if (randomRecipes) removeRandom();
    if (recipes) removeRecipes();

    // changes state of random recipes to ensure data is not appended to DOM 
    setRandomRecipes([]);
    setIsLoading(true);
    // "value" will be used to check if the query has changed
    setValue(query);
    client
      .get("complexSearch", {
        params: {
          query,
          apiKey: process.env.REACT_APP_API_KEY2,
          number: 56,
        },
      })
      .then(function (res) {
        setRecipes(res.data.results);
        setIsLoading(false);
      })
      .catch(function (error) {
        console.log(error);
      });
  };

  const getMoreRecipes = () => {
    setIsLoading(true);
    // if the value of the query has not changed, fetch more recipes
    if (query === value) {
      client
        .get("complexSearch", {
          params: {
            query,
            apiKey: process.env.REACT_APP_API_KEY4,
            number: 28,
            offset,
          },
        })
        .then(function (res) {
          setRecipes([...recipes, ...res.data.results]);
          setIsLoading(false);
          setOffset(offset + 28);
        })
        .catch(function (error) {
          console.log(error);
        });
    }
  };

  //removes recipes from DOM
  const removeRecipes = () => {
    recipes.map((recipe) => delete document.getElementById(recipe.id));
  };

  //removes random recipes from DOM 
  const removeRandom = () => {
    randomRecipes.map((recipe) => delete document.getElementById(recipe.id));
  };

  return (
    <div className="mt-5 container">
      
        <h2 className="h1 text-center mb-3">What are you in the mood for?</h2>
        <div className="row height d-flex justify-content-center align-items-center">
          <div className="col-md-8">
            <div className="search text-center">
              <i className="fa fa-search"></i>
              <input
                type="text"
                className="form-control"
                placeholder="Search by keyword"
                name="query"
                id="text-input"
                onChange={(e) => setQuery(e.target.value)}
              />
              <button
                className="search-button"
                onClick={(e) => {
                  getRecipes();
                }}
              >
                Search
              </button>
            </div>
          </div>
        </div>
      

      <div id="recipe-row" className="row">
        {isLoading && <Spinner />}
        {value !== "" && (
          <h2>{query === value ? `Search results for ${query}` : ""}</h2>
        )}

        {randomRecipes &&
          randomRecipes.map((recipe) => (
            <SmallPost
              key={recipe.id}
              id={recipe.id}
              title={recipe.title}
              img={recipe.image}
            />
          ))}
        {recipes &&
          recipes.map((recipe) => (
            <SmallPost
              key={recipe.id}
              id={recipe.id}
              title={recipe.title}
              img={recipe.image}
            />
          ))}
        {recipes.length > 0 && (
          
            <div className="text-center mb-3">
              <button
                onClick={getMoreRecipes}
                style={{ margin: "1.5rem auto", width: "125px" }}
                className="search-button"
              >
                {isLoading && recipes ? "Loading.." : "Load More"}
              </button>
            </div>
          
        )}
      </div>
    </div>
  );
};

export default SearchBar;
