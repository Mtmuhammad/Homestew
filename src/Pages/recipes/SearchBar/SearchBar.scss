@use "../../../Styles/abstracts/mixins" as *;
@use "../../../Styles/abstracts/variables" as *;

.search {
  position: relative;

  & input {
    padding: 1.5rem;
    border: 2px solid #d6d4d4;
    font-size: 1.8rem;
    border-radius: 1rem;

    &:focus {
      box-shadow: none;
      border: 2px solid $primary-green;
    }
  }

  & .fa-search {
    position: absolute;
    top: 15px;
    right: 16px;
    font-size: 3rem;
    color: $primary-green;
  }

  &-button {
    color: #fff;
    background-color: $primary-green;
    padding: 1.5rem;
    font-size: 1.5rem;
    text-transform: uppercase;
    border-radius: 5px;
    margin: 2rem 0;
    transition: all 0.3s;
    border: none;
    font-weight: 700;

    &:hover {
      color: $primary-font;
    }
  }
}
