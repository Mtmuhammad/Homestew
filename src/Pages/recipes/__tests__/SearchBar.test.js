import React from "react";
import { render, screen } from "@testing-library/react";
import SearchBar from "../SearchBar/SearchBar";

describe("smoke and snapshot tests", () => {
  test("should render without crashing", () => {
    render(<SearchBar />);
  });

  test("should match snapshot", () => {
    const { asFragment } = render(<SearchBar />);
    expect(asFragment()).toMatchSnapshot();
  });
});


test("should render input", () => {
  render(<SearchBar />);
  const input = screen.getByRole("textbox");
  expect(input).toBeInTheDocument();
  expect(input).toHaveClass("form-control");
});

test("should render button", () => {
  render(<SearchBar />);
  const button = screen.getByRole("button", { name: /search/i });
  expect(button).toBeInTheDocument();
  expect(button).toHaveClass("search-button");
});
