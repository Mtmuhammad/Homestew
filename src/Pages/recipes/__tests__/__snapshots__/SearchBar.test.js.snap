// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smoke and snapshot tests should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="mt-5 container"
  >
    <h2
      class="h1 text-center mb-3"
    >
      What are you in the mood for?
    </h2>
    <div
      class="row height d-flex justify-content-center align-items-center"
    >
      <div
        class="col-md-8"
      >
        <div
          class="search text-center"
        >
          <i
            class="fa fa-search"
          />
          <input
            class="form-control"
            id="text-input"
            name="query"
            placeholder="Search by keyword"
            type="text"
          />
          <button
            class="search-button"
          >
            Search
          </button>
        </div>
      </div>
    </div>
    <div
      class="row"
      id="recipe-row"
    >
      <div
        class="spinner-container"
      >
        <div
          class="loading-spinner"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
